### PR TITLE
feat: basic implement of `matches` fn

### PR DESCRIPTION
**Pull request by waynexia**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your intention?

The `matches` function for full-text search.

Notice this is only the first naive implementation. Some features are missing so I haven't written many docs in this PR or they will be outdated very soon. Please use the test case `evaluate_matches_without_parenthesis` as doc

Missing features:
- [ ] handle parenthesis
- [ ] validate query
- [ ] operator's priority
- [ ] must/negative/optional's conversion.

(I remember how things went with the last language project that was rushed in a week 😶‍🌫️ )

## Checklist

- [ ] I have written the necessary rustdoc comments.
- [ ] I have added the necessary unit tests and integration tests.
- [ ] This PR requires documentation updates.

## Summary by CodeRabbit

- **New Features**
  - Introduced a full-text search functionality to enhance data querying capabilities.

### CODE REVIEW

File: src/common/function/src/scalars/matches.rs [hunk 1]
```diff
@@ -0,0 +1,777 @@
...
+            match c {
+                '+' => tokens.push(Token::Must),
+                '-' => tokens.push(Token::Negative),
+                '(' => tokens.push(Token::OpenParen),
+                ')' => tokens.push(Token::CloseParen),
+                ' ' => {}
+                '\"' => {
+                    self.step_next();
+                    let phase = self.consume_next_phase(true, pattern)?;
+                    tokens.push(Token::Phase(phase));
```
Comment (zhongzc) [inline src/common/function/src/scalars/matches.rs:436]: Maybe the end here needs to consume a whitespace

File: src/common/function/src/scalars/matches.rs [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
+                        op: BinaryOp::Or,
+                        rhs: Box::new(PatternAst::Literal {
+                            op: UnaryOp::Optional,
+                            pattern: "c".to_string(),
+                        }),
+                    }),
+                },
+            ),
+            (
+                "a + b - c",
```
Comment (zhongzc) [inline src/common/function/src/scalars/matches.rs:672]: '+' '-' cannot be followed by a whitespace

**Comment by zhongzc** [comment]
I think the following BNF can be used to guide writing Parser:

```bnf
Stmt ::= OptionList MustList MustNotList

OptionList ::= ( BoolExpr )*
MustList ::= ( '+' Primary )*
MustNotList ::= ( '-' Primary )*

BoolExpr ::= AndExpr BoolExprTail
BoolExprTail ::= 'OR' AndExpr BoolExprTail | ε

AndExpr ::= Primary AndExprTail
AndExprTail ::= 'AND' Primary AndExprTail | ε

Primary ::= Term | '(' Stmt ')'
```

File: src/common/function/src/scalars/matches.rs [hunk 3]
```diff
@@ -0,0 +1,799 @@
...
+                            pattern: "b".to_string(),
+                        }),
+                    }),
+                    op: BinaryOp::Or,
+                    rhs: Box::new(PatternAst::Literal {
+                        op: UnaryOp::Must,
+                        pattern: "a".to_string(),
+                    }),
+                },
+            ),
```
Comment (WenyXu) [inline src/common/function/src/scalars/matches.rs:734]: The `lhs` and `rhs` seem to be reversed?

File: src/common/function/src/scalars/matches.rs [hunk 4]
```diff
@@ -0,0 +1,799 @@
...
+
+        ensure!(
+            !self.stack.is_empty(),
+            InvalidFuncArgsSnafu {
+                err_msg: "Empty pattern",
+            }
+        );
+
+        // conjoin them together
+        let mut builder = PatternAstBuilder::from_existing(self.stack.remove(0));
```
Comment (WenyXu) [inline src/common/function/src/scalars/matches.rs:245]: Iterate `stack` in reverse order?

File: src/common/function/src/scalars/matches.rs [hunk 5]
```diff
@@ -0,0 +1,799 @@
...
+        let result_array = result
+            .into_array(num_rows)
+            .context(GeneralDataFusionSnafu)?;
+        let result_vector =
+            BooleanVector::try_from_arrow_array(result_array).context(IntoVectorSnafu {
+                data_type: DataType::Boolean,
+            })?;
+
+        Ok(Arc::new(result_vector))
+    }
```
[the remark this answers is not in the document]
Comment (waynexia) [reply src/common/function/src/scalars/matches.rs:125]: @coderabbitai pause

**Comment by waynexia** [comment]
@coderabbitai pause
@coderabbitai resolve
@coderabbitai configuration

**Comment by waynexia** [comment]
@coderabbitai configuration

**Review by WenyXu — APPROVED** [review]

File: src/common/function/src/scalars/matches.rs [hunk 6]
```diff
@@ -0,0 +1,802 @@
...
+        DFSchema::from_unqualifed_fields(
+            [Arc::new(Field::new("data", DataType::Utf8, true))].into(),
+            HashMap::new(),
+        )
+        .unwrap()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum PatternAst {
```
Comment (killme2008) [inline src/common/function/src/scalars/matches.rs:137]: Great work! I believe we can divide this file into a separate module such as `match` or similar. The aim is to keep this file concise and organized. Take your time, it's just a suggestion.

**Pull request merged**